Starting `odc-stac` documentation

Kirill888 opened the pull request:
~For now I'm using `STAC.load` == `stacload` name on readthedocs, see here:~

~https://stacload.readthedocs.io/en/latest/index.html~
https://odc-stac.readthedocs.io/en/latest/

It's kinda sparse, but it's a start.

There is no obvious way to publish multiple projects from the same repo to readthedocs, so we might have to move docs later on when we start documenting other libs properly.

Sample notebook in this PR can be reviewed here:

https://nbviewer.jupyter.org/gist/Kirill888/4845c9f33509650e4a7a6e751d377fb6

Configuration Changes
-------------------------------

Use STAC names to describe band information
- `measurements: ` -> `assets: `
- `dtype -> data_type` to match raster extension
- `units -> unit` to match raster extension

Kirill888 committed:
Adding doc stubs for odc-stac

Kirill888 committed:
Update odc.stac.load() docs

alexgleith commented:
Is it necessary to use RST to get these happening? I find Markdown a HUGE amount easier to edit and the barrier to contributions will be much lower for others too, I think.

Kirill888 commented:
I think sphinx does support markdown for function annotations now, but I'm not sure how well or how to enable it. RST can be annoying, but it's not that bad, and it is way more precise than markdown. It is also very standard for Python documentation. For me it's easier to use RST than to deal with less supported markdown issues. VSCode has pretty good mode for it too, it's mostly just remembering to use 2 backticks instead of one. And weird url vs reference `_` positioning.

Standalone documents can be written in markdown without problem, code documentation not so much, also markdown would still have some non-standard linking mechanisms you'll need to remember, while RST is all about cross references.

Kirill888 committed:
Rename _generated/ into _api/

Kirill888 committed:
Add sample notebook for STAC.load

Kirill888 committed:
Treat documentation warnings as errors

Kirill888 committed:
Add documentation building step to main workflow

gadomski commented on libs/stac/docs/conf.py:
```diff
@@ -0,0 +1,101 @@
+
+# -- Project information -----------------------------------------------------
+
+project = "STAC.load"
```
What's the rationale behind this name, vs the PyPI name of `odc-stac`?

Kirill888 replied:
no good reason, it also lives on rtd under that name, but we should probably move it to `odc.stac` instead.

gadomski replied:
:+1: for trying to keep the same name as much as possible (`odc-stac` for urls and such, `odc.stac` for python imports).

Kirill888 replied:
```suggestion
project = "odc-stac"
```

gadomski commented on libs/stac/odc/stac/_load.py:
```diff
@@ -28,17 +28,41 @@ def load(
 
     :param items:
-       STAC Items to load
+       Collection of STAC :class:`stac.Item` to load
```
```suggestion
       Iterable of STAC :class:`stac.Item` to load
```

`Collection` is a noun with specific meaning in the STAC ecosystem.

Kirill888 replied:
Good catch, maybe `Sequence` is a more "human" name for it rather than `Iterable`? Although in Python `Sequence` is basically a kind of `Iterable` the one that supports `len()` and `[idx]`.

gadomski replied:
Since it's API documentation, IMO `Iterable` is ok to use because presumably only technical users will care about this function.

gadomski commented on libs/stac/odc/stac/_load.py:
```diff
@@ -59,23 +83,93 @@ def load(
-       Controls STAC -> Dataset conversion, mostly used to specify
-       "missing" metadata like pixel data types. see ``stac_to_ds``.
+       Controls :class:`pystac.Item` ``->`` :class:`datacube.model.Dataset`
+       conversion, mostly used to specify"missing" metadata like pixel
```
```suggestion
       conversion, mostly used to specify "missing" metadata like pixel
```

Kirill888 commented on libs/stac/docs/index.rst:
```diff
@@ -0,0 +1,28 @@
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
+STAC.load
+=========
```
```suggestion
odc-stac
========
```

Kirill888 committed:
Apply suggestions from code review

Co-authored-by: Pete Gadomski <pete.gadomski@gmail.com>

Kirill888 committed:
Fixup cross documentation linking

it's pystac.Item not stac.Item, also hides pystac

Kirill888 committed:
Using STAC names for defining missing raster extension data

For converting STAC.Item -> datacube.Dataset, use STAC raster:bands extension
names for dtype/units. Also change `measurements` section into `assets` as well.
Basic idea is that user is more familiar with STAC than with ODC so it's best to
keep user-facing names similar to STAC rather than to ODC.

Kirill888 committed:
Fix typo in commented out pip install instructions

Kirill888 committed:
Update readme, also switching it to rst

Adding
- tests and docs status
- conda installation help
- odc.stac.load example

Kirill888 committed:
Build and test is now one so link to it once

Kirill888 commented:
I have addressed comments.

I gonna merge it as I have another branch for publishing to pypi based off this one and I'm already pushing wrong commits to wrong branches.
